### DISCUSSION TITLE
Why does it take a long time to run the first time, but it takes a lot less to run the first time? How to solve the problem?

### DISCUSSION THREAD

**Discussion by xuwenlong02** [curl/curl#12718] [Q&A] [answered]
### I did this

We add the following code， but it doesn't work：

on init:
```c++
        curlMulti_ = curl_multi_init();
        if (curlMulti_ == nullptr) {
            NETSTACK_LOGE("Failed to initialize 'curl_multi'");
            return false;
        }
        curl_multi_setopt(curlMulti_, CURLMOPT_PIPELINING, CURLPIPE_MULTIPLEX);
        curlShare_ = curl_share_init();
        if (!curlShare_) {
            fprintf(stderr, "curl_share_init failed\n");
            exit(1);
        }
        curl_share_setopt(curlShare_, CURLSHOPT_SHARE, CURL_LOCK_DATA_SSL_SESSION);
        curl_share_setopt(curlShare_, CURLSHOPT_SHARE, CURL_LOCK_DATA_DNS);
        curl_share_setopt(curlShare_, CURLSHOPT_SHARE, CURL_LOCK_DATA_CONNECT);
```

on  set easy opt:
```c++
           auto ret = curl_multi_add_handle(curlMulti_, task->GetCurlHandle());
            if (ret != CURLM_OK) {
                NETSTACK_LOGE("curl_multi_add_handle err, ret = %{public}d", ret);
                StopTask(task);
            }
            curl_easy_setopt(task->GetCurlHandle(), CURLOPT_PIPEWAIT, 1L);
            if (curlShare_) {
                curl_easy_setopt(task->GetCurlHandle(), CURLOPT_SHARE, curlShare_);
            }
```
NETSTACK Debug [http_client_task.cpp 559] dnsTiming: 7.535000
NETSTACK Debug [http_client_task.cpp 560] connectTiming: 50.197000
NETSTACK Debug [http_client_task.cpp 561] tlsTiming: 261.178000
NETSTACK Debug [http_client_task.cpp 562] firstSendTiming: 261.247000
NETSTACK Debug [http_client_task.cpp 563] firstRecvTiming: 356.789000
NETSTACK Debug [http_client_task.cpp 564] totalTiming: 357.140000
NETSTACK Debug [http_client_task.cpp 565] redirectTiming: 0.000000

However, tls time sharing is not displayed.
NETSTACK Debug [http_client_task.cpp 559] dnsTiming: 0.026000
NETSTACK Debug [http_client_task.cpp 560] connectTiming: 37.605000
NETSTACK Debug [http_client_task.cpp 561] tlsTiming: 134.224000
~ DownloadCondition()
NETSTACK Debug [http_client_task.cpp 562] firstSendTiming: 134.266000
NETSTACK Debug [http_client_task.cpp 563] firstRecvTiming: 183.675000
NETSTACK Debug [http_client_task.cpp 564] totalTiming: 183.981000
NETSTACK Debug [http_client_task.cpp 565] redirectTiming: 0.000000

NETSTACK Debug [http_client_task.cpp 559] dnsTiming: 0.023000
NETSTACK Debug [http_client_task.cpp 560] connectTiming: 38.939000
NETSTACK Debug [http_client_task.cpp 561] tlsTiming: 141.003000
NETSTACK Debug [http_client_task.cpp 562] firstSendTiming: 141.058000
NETSTACK Debug [http_client_task.cpp 563] firstRecvTiming: 192.020000
NETSTACK Debug [http_client_task.cpp 564] totalTiming: 192.296000
NETSTACK Debug [http_client_task.cpp 565] redirectTiming: 0.000000

### I expected the following

The first time takes too long and the user experience is poor. How to improve the user experience?

### curl/libcurl version

curl 8.4.0

### operating system

Ubuntu 20.04

**Comment by bagder** [comment, marked as answer]
If you build the example [multi-single](https://curl.se/libcurl/c/multi-single.html) and run that with your URL, does that still run slowly? Or if you just run the curl command line tool against it?

If not, then your application has room for improvement.

If it does, then I would propose you add tracing and start figuring out where all the time is spent.
